File: sdks/csharp/PgChangeFeed.Client.Integration/PhaseEnvironment.cs
/// <summary>
/// Shared environment plumbing for the realserver integration phases
/// (ADR-0110 §Entscheidung Festlegung 2, mirrored to the C# tree): every
/// phase names its test class explicitly through the runner's
/// <c>PGCHANGEFEED_TEST_NAME</c> filter (no silent exclusion), and every
/// marker (<c>READY</c>/<c>RECEIVED</c>/<c>REJECTED</c>) is what
/// <c>tools/harness/run-sdk-csharp-integration-tests.sh</c> asserts on via
/// <c>docker logs</c> while the test still runs — so every print is flushed
/// immediately.
/// </summary>
internal static class PhaseEnvironment
{
    private static string Required(string name)
        => Environment.GetEnvironmentVariable(name)
           ?? throw new InvalidOperationException(
               $"Umgebungsvariable {name} fehlt — die Runner-Phase trägt sie.");

    /// <summary>Prints a runner marker and flushes it immediately (the
    /// runner polls <c>docker logs</c> while the test is still running).</summary>
    internal static void Print(string marker)
    {
        Console.WriteLine(marker);
        Console.Out.Flush();
    }

    internal static string Table => Required("PGCHANGEFEED_E2E_TABLE");
    internal static string Sentinel => Required("PGCHANGEFEED_E2E_SENTINEL");
    internal static string ApiToken => Required("PGCHANGEFEED_API_TOKEN");
    internal static string AdminToken => Required("PGCHANGEFEED_API_TOKEN_ADMIN");
    internal static string ReaderToken => Required("PGCHANGEFEED_API_TOKEN_READER");
    internal static string GrpcAddr => Required("PGCHANGEFEED_GRPC_ADDR");
    internal static string HttpAddr => Required("PGCHANGEFEED_HTTP_ADDR");
    internal static string NatsUrl => Required("PGCHANGEFEED_NATS_URL");
    internal static string NatsStreamToken => Required("PGCHANGEFEED_NATS_STREAM_TOKEN");
    internal static string SourceId => Required("PGCHANGEFEED_SOURCE_ID");
    internal static string HttpPublication => Required("PGCHANGEFEED_HTTP_PUBLICATION");

    internal static CancellationTokenSource ReceiveCts => new(TimeSpan.FromSeconds(90));
    internal static CancellationTokenSource RejectCts => new(TimeSpan.FromSeconds(15));
}

File: sdks/kotlin/pgchangefeed-kotlin/src/integrationTest/kotlin/io/github/pt9912/pgchangefeed/integration/PhaseEnvironment.kt
package io.github.pt9912.pgchangefeed.integration

/**
 * Shared environment plumbing for the real-server integration phases: every
 * phase names its test class explicitly through the
 * runner's `PGCHANGEFEED_TEST_NAME` `--tests` filter (no silent exclusion),
 * and every marker (`READY`/`RECEIVED`/`REJECTED`) is what
 * `tools/harness/run-sdk-kotlin-integration-tests.sh` asserts on via
 * `docker logs` while the test still runs — the Gradle task streams the
 * standard output live (`showStandardStreams`), so every print reaches the
 * runner in time.
 */
object PhaseEnvironment {
    private fun required(name: String): String =
        System.getenv(name)
            ?: throw IllegalStateException("Umgebungsvariable $name fehlt — die Runner-Phase trägt sie.")

    val table: String get() = required("PGCHANGEFEED_E2E_TABLE")
    val sentinel: String get() = required("PGCHANGEFEED_E2E_SENTINEL")
    val apiToken: String get() = required("PGCHANGEFEED_API_TOKEN")
    val adminToken: String get() = required("PGCHANGEFEED_API_TOKEN_ADMIN")
    val readerToken: String get() = required("PGCHANGEFEED_API_TOKEN_READER")
    val grpcAddr: String get() = required("PGCHANGEFEED_GRPC_ADDR")
    val httpAddr: String get() = required("PGCHANGEFEED_HTTP_ADDR")
    val natsUrl: String get() = required("PGCHANGEFEED_NATS_URL")
    val natsStreamToken: String get() = required("PGCHANGEFEED_NATS_STREAM_TOKEN")
    val sourceId: String get() = required("PGCHANGEFEED_SOURCE_ID")
    val httpPublication: String get() = required("PGCHANGEFEED_HTTP_PUBLICATION")
}